File: benchmarks/solutions/elixir/r1-word-freq.exs
case System.argv() do
  [] ->
    IO.puts(:stderr, "error: no file argument")
    System.halt(1)

  [path | _] ->
    case File.read(path) do
      {:error, reason} ->
        IO.puts(:stderr, "error: #{reason}")
        System.halt(1)

      {:ok, text} ->
        text
        |> String.downcase()
        |> String.split(~r/[^a-z]+/, trim: true)
        |> Enum.frequencies()
        |> Enum.sort_by(fn {word, count} -> {-count, word} end)
        |> Enum.take(10)
        |> Enum.each(fn {word, count} -> IO.puts("#{word}: #{count}") end)
    end
end

File: benchmarks/solutions/ruby/r1-word-freq.rb
path = ARGV[0]
unless path
  $stderr.puts "error: no file argument"
  exit 1
end

text = File.read(path)
rescue Errno::ENOENT => e
  $stderr.puts "error: #{e.message}"
  exit 1
end

counts = text.downcase.scan(/[a-z]+/).tally

counts.sort_by { |word, count| [-count, word] }
      .first(10)
      .each { |word, count| puts "#{word}: #{count}" }
